tests/components/insteon/test_config_flow.py

```diff
@@ -0,0 +1,703 @@
+MOCK_IMPORT_FULL_CONFIG_HUB_V1[CONF_X10] = [MOCK_X10_CONFIG]
+
+PATCH_CONNECTION = "homeassistant.components.insteon.config_flow.async_connect"
+PATCH_ASYNC_SETUP = "homeassistant.components.insteon.async_setup"
+PATCH_ASYNC_SETUP_ENTRY = "homeassistant.components.insteon.async_setup_entry"
+
+
+async def mock_successful_connection(*args, **kwargs):
+    """Return a successful connection."""
+    return True
+
+
+async def mock_failed_connection(*args, **kwargs):
+    """Return a failed connection."""
+    raise ConnectionError("Connection failed")
+
+
+async def _init_form(hass, modem_type):
+    """Run the init form."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+    assert result["errors"] == {}
+
+    result2 = await hass.config_entries.flow.async_configure(
+        result["flow_id"], {MODEM_TYPE: modem_type},
+    )
+    return result2
+
+
+async def _device_form(hass, flow_id, connection, user_input):
+    """Test the PLM, Hub v1 or Hub v2 form."""
+    with patch(PATCH_CONNECTION, new=connection,), patch(
+        PATCH_ASYNC_SETUP, return_value=True
+    ) as mock_setup, patch(
+        PATCH_ASYNC_SETUP_ENTRY, return_value=True,
+    ) as mock_setup_entry:
+        result = await hass.config_entries.flow.async_configure(flow_id, user_input)
+    return result, mock_setup, mock_setup_entry
+
+
+async def test_form_select_modem(hass: HomeAssistantType):
+    """Test we get a modem form."""
+    await setup.async_setup_component(hass, "persistent_notification", {})
+    result = await _init_form(hass, HUB2)
+    assert result["step_id"] == STEP_HUB_V2
+    assert result["type"] == "form"
+
+
+async def test_fail_on_existing(hass: HomeAssistantType):
+    """Test we fail if the integration is already configured."""
+    config_entry = MockConfigEntry(
+        domain=DOMAIN,
+        entry_id="abcde12345",
+        data={**MOCK_USER_INPUT_HUB_V2, CONF_HUB_VERSION: 2},
+        options={},
+    )
+    config_entry.add_to_hass(hass)
+    assert config_entry.state == config_entries.ENTRY_STATE_NOT_LOADED
+
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN,
+        data={**MOCK_USER_INPUT_HUB_V2, CONF_HUB_VERSION: 2},
+        context={"source": config_entries.SOURCE_USER},
+    )
+    assert result["type"] == data_entry_flow.RESULT_TYPE_ABORT
+    assert result["reason"] == "already_configured"
+
+
+async def test_form_select_plm(hass: HomeAssistantType):
+    """Test we set up the PLM correctly."""
+    await setup.async_setup_component(hass, "persistent_notification", {})
+    result = await _init_form(hass, PLM)
+
+    result2, mock_setup, mock_setup_entry = await _device_form(
+        hass, result["flow_id"], mock_successful_connection, MOCK_USER_INPUT_PLM
+    )
+    assert result2["type"] == "create_entry"
+    assert result2["data"] == MOCK_USER_INPUT_PLM
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5985897)